File: Data Structures & Algorithms/reorder-linked-list/submission-4.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        #find middle
        slow, fast = head, head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        #reverse half part
        second = slow.next
        pre = None
        slow.next = None
        while second:
            nxt = second.next
            second.next = pre
            pre = second
            second = nxt

        #merge
        first, second = head, pre
        while second:
            tmp1, tmp2 = first.next, second.next
            first.next = second
            second.next = tmp1
            first = tmp1
            second = tmp2
        
        

File: Data Structures & Algorithms/house-robber/submission-4.java
class Solution {
    public int rob(int[] nums) {
        int n = nums.length;
        if (n == 1) return nums[0];
        int first = nums[0];
        int second = Math.max(nums[0], nums[1]);
        for(int i = 2; i < n; i++) {
            int nxt = Math.max(first + nums[i], second);
            first = second;
            second = nxt;
        }
        return second;
    }
}
